### PR TITLE
`nix profile`: Allow referring to elements by human-readable name

### LINKED ISSUE #7960: `nix profile`s upgrade and remove have counter-intuitive argument
Using upgrade and remove is counter intuitive. The following example shows what a user will likely presume to work:

```console
$ nix profile install github:cachix/devenv/latest
$ nix profile upgrade github:cachix/devenv/latest
warning: 'github:cachix/devenv/latest' does not match any packages
warning: Use 'nix profile list' to see the current profile.
$ nix profile remove devenv
warning: 'devenv' does not match any packages
warning: Use 'nix profile list' to see the current profile.
```

- The installable used in `install` is not applicable to `remove` nor `upgrade`.
- The output shows warnings, but are actually errors. Nothing has changed to the profile.
- `nix profile list` is suggested, but it isn't clear that you need to pick the _index_ from that list.
- The name of the installable (`devenv`) is not applicable to `remove` nor `upgrade`. Ideally the `NAME` is deferred from the flake patterns (`github:OWNER/NAME/VERSION`, `github:OWNER/REPO/VERSION#NAME`, `nixpkgs#NAME`, `github:OWNER/NAME/VERSION#default`).

### PR DESCRIPTION
**Pull request by iFreilicht**
# Motivation

This PR replaces the index as the main handle for user-interaction with packages with a heuristically determined, human-readable name.

This enables commands `nix profile upgrade` and `nix profile remove` to refer to packages by their expected name like so:

```
$ nix profile install nixpkgs#vim
$ nix profile upgrade vim
$ nix profile remove vim
```

This is desirable as the index is a surprising interaction handle and the better attribute path requires matching with regex to use properly, otherwise the user gets a surprising error message. Using a name that closely matches what the user would expect is more intuitive and in line with the interface of other CLI package managers.

# Context

This PR is a starting point for improving the UX of `nix profile` in general, see #7966.
In particular, it implements the feature requested in #7967 to fix #7960, #7961 and #7962.

The old index is still displayed, but using it to `upgrade` or `remove` a package prints a deprecation warning. Actually removing the index will be done in #9171.

### Implementation strategy

A heuristic approach is used that generates the names _when the profile manifest is loaded_. The names are _not written_ to the manifest, and so its structure **does not change**, making this change 100% backward- and forward-compatible. I deem this a good approach as it allows easy additions to these heuristics should we run into edge-cases in the future.

Duplicate packages are handled by suffixing the name with a number. This is simple and effective, and I assume we don't hit that case too often in reality.

Manually overriding the auto-generated name is not supported yet, as this would only work when installing a single package at once with the current evaluation of command line arguments, which I deemed out-of-scope for this PR.

This change is compatible with profiles managed by `nix-env` as well as profiles managed by `nix profile` that still have old packages installed by `nix-env` in them.

To avoid confusing "does not match any packages" warnings, this PR also adds additional warning messages when packages _are_ found, but can't be upgraded due to having been installed via `nix-env`, a store-path or a locked flake ref:

```
warning: Found package 'vim', but it was not installed from a flake, so it can't be checked for upgrades!
warning: Found package 'github:NixOS/nixpkgs/842e90934a352f517d23963df3ec0474612e483c#legacyPackages.x86_64-linux.vim', but it was installed from a locked flake reference so it can't be upgraded!
```

this does not fix silently skipping upgrades for pinned flakes, though, see #8704.

### Alternatives

I also considered storing the generated names in the `manifest.json` but decided against it to ensure potential edge-cases in the heuristics would not be permanently stored in the profile. This will us to change the heuristics quickly without any concerns about incompatibility and users will immediately benefit from upgrades without us having to implement additional mechanisms for overriding names manually, upgrading the generated names or anything like that.

We could add the names to the output of `nix profile list --json`, and that might actually be useful, but I decided against it for now as I'm not sure we want that command's output to differ from `manifest.json`.

For deduplication of names, we might want a smarter strategy or for the names to stay consistent, which would mean storing them in the manifest, but the semantics for that interacting with `remove` are unclear, so I decided against that for now as well.

# Examples for `nix list` output:

### New profile managed entirely by `nix profile`:

```
$ outputs/out/bin/nix profile list --profile ~/repos/experiments/test-profile
Name:               vim
Index:              0
Flake attribute:    legacyPackages.x86_64-linux.vim
Original flake URL: flake:nixpkgs
Locked flake URL:   github:NixOS/nixpkgs/842e90934a352f517d23963df3ec0474612e483c
Store paths:        /nix/store/6j38m8vm8gp9a8qpw3b7dj9g50x1w95n-vim-9.0.1562

Name:               autojump
Index:              1
Flake attribute:    legacyPackages.x86_64-linux.autojump
Original flake URL: flake:nixpkgs
Locked flake URL:   github:NixOS/nixpkgs/842e90934a352f517d23963df3ec0474612e483c
Store paths:        /nix/store/aj1hslpfsw2a6vx83wmc2q8pjsafzr70-autojump-22.5.3
```

### Old profile generated by `nix-env`:

```
$ outputs/out/bin/nix profile list --profile /nix/var/nix/profiles/per-user/felix/profile-38-link
Name:               autojump
Index:              0
Store paths:        /nix/store/y7xy1cvnf1ad0by318ny6hr13dqaq6rm-autojump-22.5.3

Name:               bat
Index:              1
Store paths:        /nix/store/822la7lq7sc36m1d9l30xjl33j0iyyf1-bat-0.18.3

Name:               direnv
Index:              2
Store paths:        /nix/store/67lwbcd8wcfn0npi94dhwb16hq5mf9ab-direnv-2.28.0

Name:               fd
Index:              3
Store paths:        /nix/store/0a1xw59mzy0hwn1q7ncrmn83aamq7mkw-fd-8.2.1
```

### In a profile managed by `nix profile` with old packages installed by `nix-env` with duplicate packages:

```
outputs/out/bin/nix profile list --profile ~/repos/experiments/old-test-profile
Name:               vim
Index:              0
Store paths:        /nix/store/r37bkiiin3xfmkjm98lgpybchr9gmm15-vim-9.0.1562

Name:               asdf-vm
Index:              1
Store paths:        /nix/store/ciyg3pml5msmvq8y2m24lflkkfldq1z0-asdf-vm-0.11.3

Name:               ripgrep
Index:              2
Flake attribute:    legacyPackages.x86_64-linux.ripgrep
Original flake URL: flake:nixpkgs
Locked flake URL:   github:NixOS/nixpkgs/842e90934a352f517d23963df3ec0474612e483c
Store paths:        /nix/store/rsyhr2viynf9smk0ipma9h7lx7v1hxs3-ripgrep-13.0.0

Name:               vim1
Index:              3
Flake attribute:    legacyPackages.x86_64-linux.vim
Original flake URL: flake:nixpkgs
Locked flake URL:   github:NixOS/nixpkgs/842e90934a352f517d23963df3ec0474612e483c
Store paths:        /nix/store/6j38m8vm8gp9a8qpw3b7dj9g50x1w95n-vim-9.0.1562
```

### My own profile, managed declaratively inside a single, local flake:

```
$ outputs/out/bin/nix profile list
Name:               .dotfiles
Index:              0
Flake attribute:    packages.x86_64-linux.default
Original flake URL: path:/home/felix/.dotfiles
Locked flake URL:   path:/home/felix/.dotfiles?lastModified=1688827211&narHash=sha256-1IfyfZq3U8KDHMX2hvnPIIsHQqUUgoFcGWAw0HFgw74%3D
Store paths:        /nix/store/6rkydkpb56iavshjm2cafb341am4gcvd-ifreilicht-default-packages
```

# To-Do:

- [x] Fix tests
- [ ] Potentially fix formatting
- [x] Actually remove Index functionality
- [x] Fix `nix profile remove` not matching packages installed by `nix-env`

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by blaggacao** [comment]
>  a heuristic approach was used that generates the names when the profile manifest is loaded.

This heuristic, whatever it ends up being, is [independently useful](https://github.com/numtide/treefmt/issues/232#issuecomment-1527526665), i.e. for ecosystem commands such as `treefmt`.

**Review by bobvanderlinden — COMMENTED** [review]
This works really well!

I was a bit confused at first why `name` wasn't incorporated into `manifest.json`, but I see what you're saying. It's not clear yet if all bases are covered correctly already in terms of naming. I don't see problems with your approach atm compared to what we already have.

I caught myself attempting to make 'this could be better for UX'-type suggestions I realized this PR is a steppingstone. It does the minimal to replace indexes with names for Nix profile. This is already a step in the right direction for nix profile UX :+1: Increments on this can be done in separate PRs I think.

Thanks for the work so far already!

File: src/nix/profile.cc
Comment (bobvanderlinden) [reply src/nix/profile.cc:175]: I made some changes to have getNameFromURL in a separate file that can be unit tested. You can see the changes here: https://github.com/bobvanderlinden/nix/compare/0c4c043369445ae77c0627d39a04d9b699d11908..18d3aefc09342740e34300452b1fc88220bfd549

File: src/nix/profile-upgrade.md
Comment (iFreilicht) [reply src/nix/profile-upgrade.md:15]: Ah whoops. Fixed!

File: src/nix/profile-list.md
Comment (iFreilicht) [reply src/nix/profile-list.md:9]: This is a good point! To reduce frustration, I will not remove it in this PR and add a deprecation warning instead that refers to #9171 . This way we can get feedback without breaking anything.

File: src/nix/profile.cc
Comment (iFreilicht) [reply src/nix/profile.cc:175]: Awesome, that is indeed much easier to test! I'll incorporate this, but it does need some improvements.

Comment (iFreilicht) [reply src/nix/profile.cc]: @bobvanderlinden I cherry-picked your commits, rebased everything on master and added more test cases. I didn't get around to making them pass yet, I just went through [these docs](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-flake#url-like-syntax) and tried to find edge-cases that need to be handled.

If you have time, feel free to take a stab at it, otherwise I'll continue when I'm free again.

Comment (bobvanderlinden) [reply src/nix/profile.cc]: I have worked in this for a bit and I remember making a change to url parsing to support #. I'll have a look tomorrow.

Comment (bobvanderlinden) [reply src/nix/profile.cc]: @iFreilicht by branch includes the fixes: https://github.com/bobvanderlinden/nix/commits/profile-names-instead-of-index

https://github.com/bobvanderlinden/nix/commit/ac5874cf12ab827e569e97fbbc0bf6cd9586b90d fixes the URL parsing so that it accepts `^` (I had mistaken `#` to be the culprit, but it was `^`).

https://github.com/bobvanderlinden/nix/commit/e4e828d00d7aca93c8c34890d347795fdb743dc0 adds support for `Index` again (showing it in the `nix profile list` and referring to it using the index as argument).

https://github.com/bobvanderlinden/nix/commit/9575cddf3fe0f95892b6563c9a1d9eeae46dd600 improves on the previous regex implementation by instantiating the regex patterns once. Slight performance improvement probably.

File: src/libutil/tests/url-name.cc
Comment (bobvanderlinden) [inline src/libutil/tests/url-name.cc:57]: Nice coverage on these tests :+1:

File: src/nix/profile.cc
Comment (iFreilicht) [reply src/nix/profile.cc]: I looked at your commits now, and it doesn't seem like they help with the `getNameFromUrl` function. I'm also not sure the change regarding `#` and `^` is valid, I don't think you can install only a specific output of a flake. Or am I misunderstanding what you're trying to do here?

I already put `Index` in myself, so I skipped that one.

The last one is useful though,  I'll build upon that to support all the other schemes.

File: src/libutil/tests/url-name.cc
Comment (iFreilicht) [reply src/libutil/tests/url-name.cc:57]: Now to actually making them pass :upside_down_face:

File: src/nix/profile.cc
Comment (iFreilicht) [reply src/nix/profile.cc]: Ok I understand now what the first one was about. I incorporated that one as well and added additional tests for these kinds of URLs as well.

File: src/libutil/tests/url-name.cc
Comment (iFreilicht) [reply src/libutil/tests/url-name.cc:61]: Alright all of them are passing now!

**Comment by iFreilicht** [comment]
@bobvanderlinden if you could have another look, that would be great! I updated the documentation a bit as well. The only thing missing are the release notes now.

File: src/nix/profile.cc
Comment (TLATER) [inline src/nix/profile.cc:171]: My initial fear was that this change might cause issues if entry names match, but clearly this was considered :)

I must have misread this part of your PR message as well, which says as much:

> Duplicate packages are handled by suffixing the name with a number. This is simple and effective, and I assume we don't hit that case too often in reality.

File: src/libutil/url-name.cc
Comment (TLATER) [inline src/libutil/url-name.cc:16]: I very much appreciate the refactor between the last iteration and what I'm reading now, this is clear and understandable, even to someone less familiar with nix internals.

File: src/nix/profile.cc
Comment (TLATER) [inline src/nix/profile.cc:496]: Is this common practice in nix output? I don't think I've seen issues mentioned like this before, at least.

I understand wanting to give users agency as much as possible, but I fear this may give this issue a bit too much attention; Perhaps it's best to just let people annoyed enough by it find the issue themselves through the commit history?

File: tests/functional/nix-profile.sh
Comment (TLATER) [inline tests/functional/nix-profile.sh:137]: Could we get an integration test confirming the behavior when two flakes have matching names/heuristics?

**Review by TLATER — COMMENTED** [review]
I have one small painpoint to raise, and that is that I think users actually expect this interface:

```
$ nix profile install nixpkgs#vim
$ nix profile upgrade nixpkgs#vim
$ nix profile remove nixpkgs#vim
```

The feasibility of that is a different question, and either way this change is probably the first step to that.

@bobvanderlinden also mentions potential UX improvements they intended to suggest, and this is probably one of them, I agree this point should not block the PR at all, but I think it's worth mentioning.

Otherwise, this is great, working with names rather than numbers is obviously more human-friendly. Uniqueness is meanwhile guaranteed with numeric postfixes, so as far as I'm concerned this is a clean improvement over the current state.

It's also a beautiful diff, by the way, this is a very clean PR :)

File: src/nix/profile.cc
Comment (iFreilicht) [reply src/nix/profile.cc:496]: It is not. I wanted to raise this as an idea to allow users to participate, and gain UX feedback without having to collect metrics. Personally, I would appreciate a message like this, it makes me feel like I can be part of a discussion and decisions aren't forced on me, but I'm of course not the average user.

Maybe we can try it out and see how people react to it?

File: src/libutil/url-name.cc
Comment (iFreilicht) [reply src/libutil/url-name.cc:16]: Glad to hear it!

File: src/nix/profile.cc
Comment (iFreilicht) [reply src/nix/profile.cc:171]: To be fair, the PR message is *very* long, so I can see that details like this might be skipped over when reading it.

File: tests/functional/nix-profile.sh
Comment (iFreilicht) [reply tests/functional/nix-profile.sh:137]: Good point, I'll try to make one up :+1: The more thorough an integration test, the more potential issues it can uncover. I'd rather test a feature at multiple levels.

Comment (iFreilicht) [reply tests/functional/nix-profile.sh:144]: Ok I added one that basically copies a simple nix file to a different directory and installs it from there. Works like a charm.

**Comment by iFreilicht** [comment]
> It's also a beautiful diff, by the way, this is a very clean PR :)

Thank you, I make extensive use of `rebase -i master` and  it feels amazing.

PR is ready for another review, I now added the the release notes as well and would say this is final and ready to merge in my book.

The [one remaining discussion](https://github.com/NixOS/nix/pull/8678#discussion_r1364149072) would have to be decided by a maintainer.

**Review by TLATER — APPROVED** [review]

File: src/nix/profile.cc
Comment (fricklerhandwerk) [reply src/nix/profile.cc:496]: I think that's reasonable, but we shouldn't use the word "vote". Maybe "If this would severely impact your setup, please leave a comment explaining your use case..." or something along these lines.

Comment (bobvanderlinden) [reply src/nix/profile.cc:496]: Maybe just:

> See https://github.com/NixOS/nix/issues/9171 for more information

I really like that an URL is added :+1:, instead of guessing/searching the right place to discuss such changes further.

File: doc/manual/src/release-notes/rl-next.md
Comment (bobvanderlinden) [inline doc/manual/src/release-notes/rl-next.md:11]: _If_ deprecating indices poses a blocker for this PR I'd suggest not deprecating it for now and just add names.

File: src/libutil/tests/url-name.cc
Comment (bobvanderlinden) [inline src/libutil/tests/url-name.cc:14]: Nice going on testing the explicit outputs as well :+1:.

**Review by bobvanderlinden — APPROVED** [review]
To me this looks good :+1::

* For what it does it looks like a relatively simple change in `nix profile`
* It doesn't break existing functionality
* It doesn't change the profile manifest
* It doesn't lock us into one naming convention
* It unlocks future work on UX

The deprecation warning is fine if we can conclude that indices are indeed going to be deprecated. I'm all for it, but I can imagine it being a point of debate. I'd rather move that debate to a different PR :sweat_smile:

This implementation also made me think of ways to refactor this further. Moving part of the `getNameFromURL` logic to `src/libfetchers` (as part of `Input::fromURL`) could be useful. This will likely will trigger some discussions, so it would be great if that could be done as a follow-up PR. I'd like to look into this, but preferably not as part of this PR :sweat_smile:

I'm really hoping such refactors / iterations / improvements (and their discussions) may be done in a later stage. This PR is suited for that.

**Comment by bobvanderlinden** [comment]
@edolstra or @fricklerhandwerk could this be looked at during the next Nix team meeting?

File: src/libutil/tests/url-name.cc
Comment (iFreilicht) [reply src/libutil/tests/url-name.cc:14]: Thank you! And wouldn't you know it, adding those two tests actually uncovered a problem that had to be fixed.

File: src/nix/profile.cc
Comment (iFreilicht) [reply src/nix/profile.cc]: Yeah that's a good message. I changed it to that.

File: doc/manual/src/release-notes/rl-next.md
Comment (iFreilicht) [reply doc/manual/src/release-notes/rl-next.md:17]: Yeah I'd be fine with that.

**Comment by bobvanderlinden** [comment]
@edolstra @fricklerhandwerk I'm not sure what the system or timeline is for PRs, but this PR is in review for 3.5 months. The only feedback I've seen is that the idea has been accepted.

There are 3 people who can get behind the current implementation, but none of them have merge rights.

If there is a lack of time to review please raise this issue here or in the Nix core meetings. I'm not sure whether this happens for more PRs, if so, maybe we can look into other systematic changes to make sure there is enough time to review.

Not holding any negative thoughts, just raising an issue with the current workflow from an outside contributor's perspective.

**Comment by iFreilicht** [comment]
Rebased to fix merge conflict in release notes.

File: doc/manual/src/release-notes/rl-next.md
Comment (edolstra) [reply doc/manual/src/release-notes/rl-next.md:33]: Yeah, indices should not be removed because we need an unambiguous way to refer to profile elements.

File: src/nix/profile-list.md
Comment (edolstra) [inline src/nix/profile-list.md:39]: How do we guarantee that names are unambiguous? A name like `packages.x86_64-linux.hello` isn't.

Comment (J-Swift) [reply src/nix/profile-list.md:39]: Isnt that susceptible to the same issues this outlines: https://github.com/NixOS/nix/issues/7961?

Comment (bobvanderlinden) [reply src/nix/profile-list.md:39]: > Isnt that susceptible to the same issues this outlines: #7961?

Only for packages with the same name. #7961 is about _any_ package.

A decision about installing multiple packages with the same name can be made in a follow up PR. (Error? Replace? Interactive question? Allow duplicates?)

Comment (J-Swift) [reply src/nix/profile-list.md:39]: Maybe it makes sense to use some portion of the store hash as an alternative suffix? e.g. hello-3bjqzx, hello-6r9h8

Comment (bobvanderlinden) [reply src/nix/profile-list.md:39]: That could also be an option.

Personally I'd like to avoid duplicate names eventually altogether by erroring upon install. I think these kinds of improvements/iterations should be a decision for a follow-up PR. The current implementation of this PR is simple and more or less in line with the current implementation on master (numbering the packages).

I see this PR as the most minimal step we can make towards using names for profile entries.

**Comment by bobvanderlinden** [comment]
@edolstra @fricklerhandwerk could this get another look? What does this need to get approval/rejection? Or what does this need to speed up the reviewing process?

**Comment by szlend** [comment]
It kind of feels like nobody in the Nix team actually uses or cares about `nix profile`. I've been trying to advocate Nix in our company and I'm frankly embarrassed every time I have to explain to someone how to upgrade or remove a package. This PR is a small backwards compatible change and a massive UX improvement and I wish it was taken more seriously and not just left here to rot without any feedback.

**Comment by fricklerhandwerk** [comment]
@edolstra said to take a look this week

**Comment by bobvanderlinden** [comment]
🥳  thanks for the review + merge!

File: src/libutil/url-name.hh
Comment (Ericson2314) [inline src/libutil/url-name.hh:20]: @iFreilicht I would appreciate if you could move this elsewhere. `libutil` should be just utilities and not define Nix-specific concepts like store paths or flake things. (In the "Domain-Driven Design" parlance, this is a "application layer" or "domain layer" concept, but libutil is supposed to be the "infrastructure layer".)

Comment (iFreilicht) [reply src/libutil/url-name.hh:20]: Sure. Could you maybe give me a hint where "elsewhere" could be?

Comment (Ericson2314) [reply src/libutil/url-name.hh:20]: Sure. I would put in `src/libexpr/flake/` for now. There are other flakes things in there that don't strictly relate to evaluation already.

Comment (iFreilicht) [reply src/libutil/url-name.hh:20]: Done, see #9655

**Pull request merged**
